cmd/checkpoint/README.md

```diff
@@ -1,16 +1,74 @@
 # Checkpoint
 
-`checkpoint` is a small utility that will convert an API server pod manifest
-into a static manifest which can be stored on disk. The pod manifest is
-obtained from the kubelets read only API.
-
-This is useful for single-master clusters. A single-master cluster with this
-checkpointing is resilient to certain failure scenarios, such as a system
-reboot, or complete loss of the API server.
-
-The tool waits until it detects an API server is no longer running.
-In that case, it will move the static pod manifest created earlier into the
-directory specified as the kubelets config directory. Once the tool detects our
-normal API server has started up again, it will move the static manifest out of
-the config directory, causing the kubelet to stop the checkpointed API server,
-enabling the self-hosted API server to take over once more.
+## Description
+
+`checkpoint` is a utility application which will manage "checkpoints" of pods scheduled to the local node.
+
+The purpose of a pod checkpoint is to ensure that existing local pod state can be recovered in the absence of an api-server.
+
+The kubelet will already attempt to ensure that local pods will continue to run in the absence of an API server (based on restartPolicy). However, if all local runtime state has been lost (for example after a reboot), the checkpoint utility will ensure the local state can be recovered until an api-server is contacted.
+
+This is accomplished by managing checkpoints as static pod manifests:
+
+- When the checkpointer sees that a "parent pod" (a pod which should be checkpointed), is succesfully running, the checkpointer will save a local copy of the manifest.
+
+- If the parent pod is detected as no longer running, the checkpointer will "activate" the checkpoint manifest. It will allow the checkpoint to continue running until the parent-pod is restarted on the local node, or it is able to contact an api-server to determine that the parent pod is no longer scheduled to this node.
+
+## Use
+
+Any pod which contains the `checkpointer.alpha.coreos.com/checkpoint=true` annotation will be considered a viable "parent pod" which should be checkpointed.
+The parent pod cannot itself be a static pod, and is not a checkpoint itself.
+
+Checkpoints are denoted by the `checkpointer.alpha.coreos.com/checkpoint-of` annotation. This annotation will point to the parent of this checkpoint, and will be the `namespace/name` of the pod.
+
+For example the pod:
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: kube-apiserver
+  namespace: kube-system
+  annotations:
+    checkpointer.alpha.coreos.com/checkpoint=true
+```
+
+Will generate a checkpointed pod as:
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: kube-apiserver
+  namespace: kube-system
+  annotations:
+    checkpointer.alpha.coreos.com/checkpoint-of=kube-system/kube-apiserver
```

Comment:
not sure we need to append kube-system in this annotation (we might need it for the on disk file name). we already have namespace in metadata.